File: concepts/03_errors_and_resources/01_resource_cleanup/r/test_02_partial_acquisition_and_control_flow.R
# polyglot-family: errors_and_resources
# polyglot-concept: resource_cleanup
# polyglot-related: languages/r/language/11_conditions_control_flow_and_cleanup/
# polyglot-related+: test_086_on_exit_multiple_cleanup_actions.R
#
# 共同问题：部分取得资源后失败怎样回滚；return 是否绕过清理。
# 对照观察：只在成功取得后登记 `on.exit`；显式 return 与错误都会先执行已登记动作。

trace <- character()
acquire <- function(fail) {
    trace <<- c(trace, "first")
    on.exit(trace <<- c(trace, "release-first"), add = TRUE)
    if (fail) stop("second failed")
    trace <<- c(trace, "second")
    on.exit(trace <<- c(trace, "release-second"), add = TRUE)
    return("done")
}

stopifnot(
    inherits(tryCatch(acquire(TRUE), error = identity), "error"),
    identical(trace, c("first", "release-first"))
)

File: concepts/03_errors_and_resources/01_resource_cleanup/lua/test_02_partial_acquisition_and_control_flow.lua
-- Common question: which resources are cleaned when acquisition or control flow stops early?
-- Inputs: two successful acquisitions, one failed acquisition, return, and error.
-- Observations: declaration timing, reverse close order, partial ownership, and nil/false close values.
-- polyglot-family: errors_and_resources
-- polyglot-concept: resource_cleanup
-- polyglot-related: languages/lua/language/07_errors_and_resources/test_052_to_be_closed_normal_exit.lua

local t = require("support.assertions")
local events = {}
local function acquire(name)
    if name == "failed" then
        return nil, "not acquired"
    end
    return setmetatable({name = name}, {
        __close = function(self)
            events[#events + 1] = self.name
        end,
    })
end

local ok = pcall(function()
    local first <close> = assert(acquire("first"))
    local second <close> = assert(acquire("second"))
    local third <close> = assert(acquire("failed"))
    return first, second, third
end)
t.falsey(ok)
t.equal(table.concat(events, ","), "second,first")

do
    local ignored <close> = nil
    t.equal(ignored, nil)
end

t.done()
